# Uniformly Random Knapsack Packings

A binary knapsack problem involves a vector of positive weights, w=(w<sub>1</sub>,...,w<sub>n</sub>), and a maximum weight, W. A *feasible solution* or *packing* consists of a binary vector, a=(a<sub>1</sub>,...,a<sub>n</sub>), such that the dot product of a with w does not exceed W.

A knapsack problem per se generally assigns values to items and the problem is to maximize cumulative value subject to the constraint on weight. Such a result is called an *optimal* feasible solution or packing. Values are irrelevant to the objective here, however, which is to generate uniformly random knapsack packings which satisfy the weight constraint. 

We use the MCMC algorithm given in B Morris, A Sinclair, Random walks on truncated cubes and sampling 0-1 knapsack solutions SIAM journal on computing, 2004 - SIAM. [Google Scholar](https://scholar.google.com/scholar?q=RANDOM+WALKS+ON+TRUNCATED+CUBES+AND+SAMPLING+0-1+KNAPSACK+SOLUTIONS&btnG=&hl=en&as_sdt=0%2C21). It is described as follows. Given a current packing:

* With probability 1/2, do nothing. (This is a technicality to avoid Markov Chain periodicity.)
* Pick an item, i, at random.
* If i is in the current packing, take it out.
* If i is not in the current packing, and adding it does not cause the weight limit to be exceeded, put it in.

The reference proves that the chain converges rapidly to a uniform sampler.

The following code requires the `Mamba`, `Plots`, and `PyPlot` packages in addition to `HitAndRun`.


#### 100 loaves of bread

Weights, whether of persons or objects, are often distributed according to normal distributions (restricted to the positive reals.) Weights of hand-baked, 1 lb. loaves of bread, for example, may have a standard deviation of 0.2 lb. Assume 100 such loaves are baked. Simulating the result:

In [1]:
w = zeros(Float64,100);
srand(1234);
while !all(w .> 0.0)
    w = 1.0 + 0.2*randn(100);
end
using Plots
pyplot();
histogram(w);
xlabel!("Weight (lbs.)");
ylabel!("Count")
title!("Weight Distribution of 100 Loaves")

[Plots.jl] Initializing backend: pyplot


<?xml version="1.0" encoding="utf-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN"
 "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<!-- Created with matplotlib (http://matplotlib.org/) -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 0.2 -->
 
 
 <path d="
M31.7812 66.4062
Q24.1719 66.4062 20.3281 58.9062
Q16.5 51.4219 16.5 36.375
Q16.5 21.3906 20.3281 13.8906
Q24.1719 6.39062 31.7812 6.39062
Q39.4531 6.39062 43.2812 13.8906
Q47.125 21.3906 47.125 36.375
Q47.125 51.4219 43.2812 58.9062
Q39.4531 66.4062 31.7812 66.4062
M31.7812 74.2188
Q44.0469 74.2188 50.5156 64.5156
Q56.9844 54.8281 56.9844 36.375
Q56.9844 17.9688 50.5156 8.26562
Q44.0469 -1.42188 31.7812 -1.42188
Q19.5312 -1.42188 13.0625 8.26562
Q6.59375 17.9688 6.59375 36.375
Q6.59375 54.8281 13.0625 64.5156
Q19.5312 74.2188 31.7812 74.2188" id="DejaVuSans-30"/>
 <path d="
M19.1875 8.29688
L53.6094 8.29688
L53.6094 0
L7.32812 0
L7.32812 8.29688
Q12.9375 14.1094 22.625 23.8906
Q32.3281 33.6875 34.8125 36.5312
Q39.5469 41.8438 41.4219 45.5312
Q43.3125 49.2188 43.3125 52.7812
Q43.3125 58.5938 39.2344 62.25
Q35.1562 65.9219 28.6094 65.9219
Q23.9688 65.9219 18.8125 64.3125
Q13.6719 62.7031 7.8125 59.4219
L7.8125 69.3906
Q13.7656 71.7812 18.9375 73
Q24.125 74.2188 28.4219 74.2188
Q39.75 74.2188 46.4844 68.5469
Q53.2188 62.8906 53.2188 53.4219
Q53.2188 48.9219 51.5312 44.8906
Q49.8594 40.875 45.4062 35.4062
Q44.1875 33.9844 37.6406 27.2188
Q31.1094 20.4531 19.1875 8.29688" id="DejaVuSans-32"/>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 0.4 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 0.6 -->
 
 <path d="
M33.0156 40.375
Q26.375 40.375 22.4844 35.8281
Q18.6094 31.2969 18.6094 23.3906
Q18.6094 15.5312 22.4844 10.9531
Q26.375 6.39062 33.0156 6.39062
Q39.6562 6.39062 43.5312 10.9531
Q47.4062 15.5312 47.4062 23.3906
Q47.4062 31.2969 43.5312 35.8281
Q39.6562 40.375 33.0156 40.375
M52.5938 71.2969
L52.5938 62.3125
Q48.875 64.0625 45.0938 64.9844
Q41.3125 65.9219 37.5938 65.9219
Q27.8281 65.9219 22.6719 59.3281
Q17.5312 52.7344 16.7969 39.4062
Q19.6719 43.6562 24.0156 45.9219
Q28.375 48.1875 33.5938 48.1875
Q44.5781 48.1875 50.9531 41.5156
Q57.3281 34.8594 57.3281 23.3906
Q57.3281 12.1562 50.6875 5.35938
Q44.0469 -1.42188 33.0156 -1.42188
Q20.3594 -1.42188 13.6719 8.26562
Q6.98438 17.9688 6.98438 36.375
Q6.98438 53.6562 15.1875 63.9375
Q23.3906 74.2188 37.2031 74.2188
Q40.9219 74.2188 44.7031 73.4844
Q48.4844 72.75 52.5938 71.2969" id="DejaVuSans-36"/>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 0.8 -->
 
 <path d="
M31.7812 34.625
Q24.75 34.625 20.7188 30.8594
Q16.7031 27.0938 16.7031 20.5156
Q16.7031 13.9219 20.7188 10.1562
Q24.75 6.39062 31.7812 6.39062
Q38.8125 6.39062 42.8594 10.1719
Q46.9219 13.9688 46.9219 20.5156
Q46.9219 27.0938 42.8906 30.8594
Q38.875 34.625 31.7812 34.625
M21.9219 38.8125
Q15.5781 40.375 12.0312 44.7188
Q8.5 49.0781 8.5 55.3281
Q8.5 64.0625 14.7188 69.1406
Q20.9531 74.2188 31.7812 74.2188
Q42.6719 74.2188 48.875 69.1406
Q55.0781 64.0625 55.0781 55.3281
Q55.0781 49.0781 51.5312 44.7188
Q48 40.375 41.7031 38.8125
Q48.8281 37.1562 52.7969 32.3125
Q56.7812 27.4844 56.7812 20.5156
Q56.7812 9.90625 50.3125 4.23438
Q43.8438 -1.42188 31.7812 -1.42188
Q19.7344 -1.42188 13.25 4.23438
Q6.78125 9.90625 6.78125 20.5156
Q6.78125 27.4844 10.7812 32.3125
Q14.7969 37.1562 21.9219 38.8125
M18.3125 54.3906
Q18.3125 48.7344 21.8438 45.5625
Q25.3906 42.3906 31.7812 42.3906
Q38.1406 42.3906 41.7188 45.5625
Q45.3125 48.7344 45.3125 54.3906
Q45.3125 60.0625 41.7188 63.2344
Q38.1406 66.4062 31.7812 66.4062
Q25.3906 66.4062 21.8438 63.2344
Q18.3125 60.0625 18.3125 54.3906" id="DejaVuSans-38"/>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 1.0 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 1.2 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 1.4 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 1.6 -->
 
 
 
 
 
 
 
 
 <!-- Weight (lbs.) -->
 
 
 <path d="
M44.2812 53.0781
L4

#### Knapsacks with a 50 lb. limit

Suppose that a typical courier can carry at most 50 lbs. of bread. With the above population of loaves, what are the characteristics of a random (not necessarily optimal) load? We generate 100,000 loads, skipping the first 1000 to allow for convergence to uniform sampling, and we thin by 50 thereafter.

In [2]:
include("../src/HitAndRun.jl") # skip this line if HitAndRun is installed as a package
using HitAndRun
W = 50.0; # 50 lb. weight limit
packings = mcmc_knapsack(w, W, 1000, 50, 100000);

Cumulative weights are in the first column of `packings`, the packings themselves are 1-0 vectors in the following 100 columns. We apply Mamba's Geweke diagnostic to weights to assess convergence.

In [3]:
using Mamba;
# Convert weights to an object of type Mamba.Chains
chain = hitrun2chains(packings[:,1], 1000, 50);
# Apply diagnostic
gewekediag(chain)

  Z-score p-value
T    1.09  0.2755



The p-value, 0.2755, is consistent with the diagnostic's null hypothesis, namely that convergence has occurred. Thus the packings are approximately random.

In may be of interest to see how the number of loaves varies with the cumulative weight of packings.

In [8]:
# Count the loaves in each sample
loaves = sum(packings[:,2:101],2);
# Scatterplot
scatter(packings[:,1], loaves)
title!("Loaves vs. Weight")
xlabel!("Weight (lbs.)")
ylabel!("Number of loaves")

<?xml version="1.0" encoding="utf-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN"
 "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<!-- Created with matplotlib (http://matplotlib.org/) -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 25 -->
 
 <path d="
M19.1875 8.29688
L53.6094 8.29688
L53.6094 0
L7.32812 0
L7.32812 8.29688
Q12.9375 14.1094 22.625 23.8906
Q32.3281 33.6875 34.8125 36.5312
Q39.5469 41.8438 41.4219 45.5312
Q43.3125 49.2188 43.3125 52.7812
Q43.3125 58.5938 39.2344 62.25
Q35.1562 65.9219 28.6094 65.9219
Q23.9688 65.9219 18.8125 64.3125
Q13.6719 62.7031 7.8125 59.4219
L7.8125 69.3906
Q13.7656 71.7812 18.9375 73
Q24.125 74.2188 28.4219 74.2188
Q39.75 74.2188 46.4844 68.5469
Q53.2188 62.8906 53.2188 53.4219
Q53.2188 48.9219 51.5312 44.8906
Q49.8594 40.875 45.4062 35.4062
Q44.1875 33.9844 37.6406 27.2188
Q31.1094 20.4531 19.1875 8.29688" id="DejaVuSans-32"/>
 <path d="
M10.7969 72.9062
L49.5156 72.9062
L49.5156 64.5938
L19.8281 64.5938
L19.8281 46.7344
Q21.9688 47.4688 24.1094 47.8281
Q26.2656 48.1875 28.4219 48.1875
Q40.625 48.1875 47.75 41.5
Q54.8906 34.8125 54.8906 23.3906
Q54.8906 11.625 47.5625 5.09375
Q40.2344 -1.42188 26.9062 -1.42188
Q22.3125 -1.42188 17.5469 -0.640625
Q12.7969 0.140625 7.71875 1.70312
L7.71875 11.625
Q12.1094 9.23438 16.7969 8.0625
Q21.4844 6.89062 26.7031 6.89062
Q35.1562 6.89062 40.0781 11.3281
Q45.0156 15.7656 45.0156 23.3906
Q45.0156 31 40.0781 35.4375
Q35.1562 39.8906 26.7031 39.8906
Q22.75 39.8906 18.8125 39.0156
Q14.8906 38.1406 10.7969 36.2812
z
" id="DejaVuSans-35"/>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 30 -->
 
 <path d="
M40.5781 39.3125
Q47.6562 37.7969 51.625 33
Q55.6094 28.2188 55.6094 21.1875
Q55.6094 10.4062 48.1875 4.48438
Q40.7656 -1.42188 27.0938 -1.42188
Q22.5156 -1.42188 17.6562 -0.515625
Q12.7969 0.390625 7.625 2.20312
L7.625 11.7188
Q11.7188 9.32812 16.5938 8.10938
Q21.4844 6.89062 26.8125 6.89062
Q36.0781 6.89062 40.9375 10.5469
Q45.7969 14.2031 45.7969 21.1875
Q45.7969 27.6406 41.2812 31.2656
Q36.7656 34.9062 28.7188 34.9062
L20.2188 34.9062
L20.2188 43.0156
L29.1094 43.0156
Q36.375 43.0156 40.2344 45.9219
Q44.0938 48.8281 44.0938 54.2969
Q44.0938 59.9062 40.1094 62.9062
Q36.1406 65.9219 28.7188 65.9219
Q24.6562 65.9219 20.0156 65.0312
Q15.375 64.1562 9.8125 62.3125
L9.8125 71.0938
Q15.4375 72.6562 20.3438 73.4375
Q25.25 74.2188 29.5938 74.2188
Q40.8281 74.2188 47.3594 69.1094
Q53.9062 64.0156 53.9062 55.3281
Q53.9062 49.2656 50.4375 45.0938
Q46.9688 40.9219 40.5781 39.3125" id="DejaVuSans-33"/>
 <path d="
M31.7812 66.4062
Q24.1719 66.4062 20.3281 58.9062
Q16.5 51.4219 16.5 36.375
Q16.5 21.3906 20.3281 13.8906
Q24.1719 6.39062 31.7812 6.39062
Q39.4531 6.39062 43.2812 13.8906
Q47.125 21.3906 47.125 36.375
Q47.125 51.4219 43.2812 58.9062
Q39.4531 66.4062 31.7812 66.4062
M31.7812 74.2188
Q44.0469 74.2188 50.5156 64.5156
Q56.9844 54.8281 56.9844 36.375
Q56.9844 17.9688 50.5156 8.26562
Q44.0469 -1.42188 31.7812 -1.42188
Q19.5312 -1.42188 13.0625 8.26562
Q6.59375 17.9688 6.59375 36.375
Q6.59375 54.8281 13.0625 64.5156
Q19.5312 74.2188 31.7812 74.2188" id="DejaVuSans-30"/>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 35 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 40 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 45 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 50 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 <!-- 55 -->
 
 
 
 
 
 
 
 <!-- Weight (lbs.) -->
 
 
 <path d="
M44.2812 53.0781
L44.2812 44.5781
Q40.4844 46.5312 36.375 47.5
Q32.2812 48.4844 27.875 48.4844
Q21.1875 48.4844 17.8438 46.4375
Q14.5 44.3906 14.5 40.2812
Q14.5 37.1562 16.8906 35.375
Q19.2812 33.5938 26.5156 31.9844
L29.5938 31.2969
Q39.1562 29.25 43.1875 25.5156
Q47.2188 21.7812 47.2188 15.0938
Q47.2188 7.46875 41.1875 3.01562
Q35.1562 -1.42188 24.6094 -1.42188
Q20.2188 -1.42188 15.4531 -0.5625
Q10.6875 0.296875 5.42188 2
L5.42188 11.2812
Q10.4062 8.6875 15.2344 7.39062
Q20.0625 6.10938 24.8125 6.10938
Q31.156